### PR TITLE
convert FETCH_COOKIES requests to use curl for better authentication support

### PR DESCRIPTION
**Pull request by gtrout**
The original file_get_contents method wasn't passing all the cookies needed to render private/authenticated content. I tried converting it to use curl and it worked, so here are my changes for your consideration.

Credit to stackoverflow user **PiTheNumber** for the curl example I started with (http://stackoverflow.com/a/23907823/3735757).

### CODE REVIEW

File: wp-pdf-templates.php [hunk 1]
```diff
@@ -207,24 +207,47 @@ function _use_pdf_template() {
...
+		
+		// store the cookies
+		$cookies = array();
+		foreach ($_COOKIE as $key => $value)
+		{
+			if ($key != 'Array')
+			{
+				$cookies[] = $key . '=' . $value;
+			}
+		}
```
Comment (anttiviljami) [inline wp-pdf-templates.php]: Why loop the cookies and reconstruct them to what's essentially $_SERVER['HTTP_COOKIE'] ?

**Comment by anttiviljami** [comment]
Thanks for the PR! I'll look into this and test it on a couple of sites that use FETCH_COOKIES.

File: wp-pdf-templates.php [hunk 2]
```diff
@@ -207,24 +207,47 @@ function _use_pdf_template() {
...
+		curl_setopt( $ch, CURLOPT_COOKIE, implode(';', $cookies) );
+		
+		// Stop session so curl can use the same session without conflicts
+		session_write_close();
+		
+		$response = curl_exec($ch);
+		curl_close($ch);
+		
+		// Session restart
+		session_start();
```
Comment (anttiviljami) [inline wp-pdf-templates.php]: What if there's no pre-existing session? We really don't want to start an unnecessary session for users that just want to download PDFs.

File: wp-pdf-templates.php [hunk 3]
```diff
@@ -207,24 +207,47 @@ function _use_pdf_template() {
...
+		
+		// Seperate header and body
+		list($header, $body) = explode("\r\n\r\n", $response, 2);
```
Comment (anttiviljami) [inline wp-pdf-templates.php]: This seems wrong. Instead, we should do something like:

```
$header_len = curl_getinfo( $ch, CURLINFO_HEADER_SIZE );
$header = substr( $response, 0, $header_len );
$body = substr( $response, $header_len );
```

File: wp-pdf-templates.php [hunk 4]
```diff
@@ -207,24 +207,47 @@ function _use_pdf_template() {
...
+		
+		// get http headers for cookies
+		curl_setopt($ch, CURLOPT_VERBOSE, 1);
+		curl_setopt($ch, CURLOPT_HEADER, 1);
+		
+		// store the cookies
+		$cookies = array();
+		foreach ($_COOKIE as $key => $value)
+		{
+			if ($key != 'Array')
```
Comment (anttiviljami) [inline wp-pdf-templates.php]: Why is this here? Comment the reason for the exclusion please.

**Comment by anttiviljami** [comment]
Could you also fix the code style to match WordPress.org PHP Coding standards?

https://make.wordpress.org/core/handbook/coding-standards/php/

... with the exception of the indentation tab rule. Tabs should be 2 spaces for better readability, thanks ! :)

File: wp-pdf-templates.php (hunk 1 above)
Comment (gtrout) [inline wp-pdf-templates.php]: I actually think this is what fixed my original problem of all cookies not being sent with the original file_get_contents method. This guarantees that multiple cookies are formatted in the way that the cURL session expects per: http://php.net/manual/en/function.curl-setopt.php

Comment (anttiviljami) [inline wp-pdf-templates.php]: That's odd. Could you give me an example of where this method would give a different result from $_SERVER['HTTP_COOKIE']? To my understanding, $_SERVER['HTTP_COOKIE'] should already contain the exact representation that libcurl uses anyways.

**Comment by gtrout** [comment]
I've pushed a commit that detabs the code.

**Comment by anttiviljami** [comment]
Thanks ! Can you just fix https://make.wordpress.org/core/handbook/coding-standards/php/#brace-style as well?

**Comment by gtrout** [comment]
Definitely! Hanging braces like that are fugly. Fixed.

File: wp-pdf-templates.php (hunk 2 above)
Comment (gtrout) [inline wp-pdf-templates.php]: Isn't this closing and reopening the existing session (opened by `curl_init()`) so that the curl doesn't require an additional one?

File: wp-pdf-templates.php (hunk 3 above)
Comment (gtrout) [inline wp-pdf-templates.php]: Your way may be cleaner, sure.

File: wp-pdf-templates.php (hunk 2 above)
Comment (anttiviljami) [inline wp-pdf-templates.php]: curl_init() doesn't open a PHP session, like session_start(). It just starts something called a curl session, which is different from a PHP session with cookies and sessionfiles. 

So.. opening and closing a php session is unnecessary?

File: wp-pdf-templates.php (hunk 4 above)
Comment (gtrout) [inline wp-pdf-templates.php]: I guess it's to keep serialized values out of our new array?

Comment (anttiviljami) [inline wp-pdf-templates.php]: Surely we can't just drop cookies with serialised values?

File: wp-pdf-templates.php (hunk 2 above)
Comment (gtrout) [inline wp-pdf-templates.php]: I guess that may be the case, since we're only performing one operation with this session. If we were doing more than one thing we'd want to unlock the session with session_write_close(). See: http://blog.psych0tik.net/2012/12/a-look-at-php-session-write-locks/comment-page-1/

File: wp-pdf-templates.php (hunk 4 above)
Comment (gtrout) [inline wp-pdf-templates.php]: I guess the foreach isn't recursive? Maybe it should be something like:

```
array_walk_recursive($_COOKIE, function ($value, $key) {
  $cookies[] = $key . '=' . $value;
});
```

File: wp-pdf-templates.php (hunk 2 above)
Comment (anttiviljami) [inline wp-pdf-templates.php]: Right! Now I understand. Because curl invokes a new php heap that possibly accesses the same sessionfile, we need to suspend the session while executing the curl request. That makes sense.

However, this should only be done when a previous session exists. Otherwise session_start() on line 239 would just start a new session for no reason whatsoever which is really bad. Can you fix this by checking if there's a session and then deciding whether to open and close it again?

File: wp-pdf-templates.php (hunk 4 above)
Comment (anttiviljami) [inline wp-pdf-templates.php]: Maybe, but I'm still not convinced we should be looping the $_COOKIE array at all. Why does $_SERVER['HTTP_COOKIE'] not work here?

File: wp-pdf-templates.php (hunk 3 above)
Comment (anttiviljami) [inline wp-pdf-templates.php]: Also, I don't think that explode("\r\n\r\n" would work reliably enough. What if the web server doesn't use carriage returns but just newlines?

**Comment by anttiviljami** [comment]
@gtrout: Any updates regarding my previous questions?

**Comment by gtrout** [comment]
Sorry, haven't had time to get back to this. I will try to make some changes in the next week or so.
